app/controllers/AddBusinessTradeController.scala

```diff
@@ -0,0 +1,131 @@
+                                           checkSessionTimeout: SessionTimeoutPredicate,
+                                           retrieveNino: NinoPredicate,
+                                           val addBusinessTradeView: AddBusinessTrade,
+                                           val retrieveIncomeSources: IncomeSourceDetailsPredicate,
+                                           val retrieveBtaNavBar: NavBarPredicate,
+                                           val itvcErrorHandler: ItvcErrorHandler,
+                                           incomeSourceDetailsService: IncomeSourceDetailsService)
+                                          (implicit val appConfig: FrontendAppConfig,
+                                             implicit val itvcErrorHandlerAgent: AgentItvcErrorHandler,
+                                             implicit override val mcc: MessagesControllerComponents,
+                                             val ec: ExecutionContext)
+  extends ClientConfirmedController with I18nSupport with FeatureSwitching{
+
+  val backURL: String = "/report-quarterly/income-and-expenses/view/income-sources/add/business-start-date-check"
+  val agentBackURL: String = "/report-quarterly/income-and-expenses/view/agents/income-sources/add/business-start-date-check"
+  def show: Action[AnyContent] = (checkSessionTimeout andThen authenticate andThen retrieveNino
+    andThen retrieveIncomeSources andThen retrieveBtaNavBar).async {
+    implicit user =>
+      handleRequest(isAgent = false)
+  }
+
+  def showAgent: Action[AnyContent] =
+    Authenticated.async {
+      implicit request =>
+        implicit user =>
+          getMtdItUserWithIncomeSources(incomeSourceDetailsService, useCache = true) flatMap {
+            implicit mtdItUser =>
+              handleRequest(isAgent = true)
+          }
+    }
+
+  def handleRequest(isAgent: Boolean)(implicit user: MtdItUser[_], ec: ExecutionContext): Future[Result] = {
+    Future {
+      if (isDisabled(IncomeSources)) {
+        Redirect(controllers.routes.HomeController.show())
+      } else {
+        if(!isAgent) Ok(addBusinessTradeView(BusinessTradeForm.form, routes.AddBusinessTradeController.submit(), isAgent, backURL, agentBackURL, false))
+        else Ok(addBusinessTradeView(BusinessTradeForm.form, routes.AddBusinessTradeController.agentSubmit(), isAgent, backURL, agentBackURL, false))
+      }
+    }
+  }
+
+  def submit: Action[AnyContent] = (checkSessionTimeout andThen authenticate andThen retrieveNino
+    andThen retrieveIncomeSources andThen retrieveBtaNavBar).async {
+    implicit request =>
+      BusinessTradeForm.form.bindFromRequest().fold(
+        formWithErrors => {
+          Future {
+            Ok(addBusinessTradeView(formWithErrors, routes.AddBusinessTradeController.submit(), false, backURL, agentBackURL, false))
+          }
+        },
+        formData => {
+          if (formData.trade == request.session.get("addBusinessName").getOrElse("????")){
```

Comment:
I am not sure what we are trying to do by reading "addBusinessName" from session storage?

Reply:
One of the verification constraints is that the trade cannot be the same as the business name, so I compare them here

Reply:
I couldn't find an easy way to add this as a constraint in the Form, which means there is also some strange code in the view to make this work (and show the error correctly)

Reply:
The "????" is what to replace the name with if there is none in the session storage, which will only happen in the unit tests as in the user flow they have to have input a valid business name to get to this page

Reply:
This method is called by every test, so I need either a way to add name to session storage in the test, or a failsafe in case there isn't any. I'll work on adding a name from the test spec

Reply:
* for the "happy path" scenarios you should be able to add required data to request session storage within a test;
* for failure case (when there is no required data in session storage) you should expect relevant failures etc;